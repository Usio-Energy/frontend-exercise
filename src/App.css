.App {
  text-align: center;
}

.App-logo {
  height: 4rem;
}

.App-header {
  background-color: #fff;
  padding: 1.5rem;
  color: #333;
}

.App-title {
  font-size: 1.5em;
}

.App-intro {
  font-size: large;
}