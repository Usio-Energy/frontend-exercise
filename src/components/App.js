import React from 'react';

const App = () => (
  <div>
    <h1>To do app</h1>

    <input type="text" /> <button>Add</button>

    <div className="status">10 out of 20 tasks completed</div>

    <ul>
      <li>
        <input type="checkbox" /> Item 1
      </li>

      <li>
        <input type="checkbox" /> Item 2
      </li>

      <li>
        <input type="checkbox" /> Item 3
      </li>
    </ul>
  </div>
);

export default App;