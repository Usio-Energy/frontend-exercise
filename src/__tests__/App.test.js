import React from 'react';
import { shallow } from 'enzyme';

import App from '../components/App';

describe('App', () => {
  let app;

  beforeAll(() => {
    app = shallow(<App />);

    app.setState({
      items: [
        {
          id: 1,
          value: 'Item 1',
          isComplete: false,
        },
        {
          id: 2,
          value: 'Item 2',
          isComplete: false,
        }
      ]
    })
  });

  describe('Render', () => {
    it('should render correctly', () => {
      expect(app.getElement()).toMatchSnapshot();
    });
  })

  describe('Tasks', () => {
    it('should render items from state', () => {
      expect(app.find('Item')).toHaveLength(app.state().items.length);
    });

    it('should mark an item as complete when it\'s checkbox is checked', () => {
      app.instance().onStatusChange(2, true);

      expect(app.state().items[1].isComplete).toEqual(true);
    });

    it('should pass `onStatusChanged` as prop to all <Item /> components', () => {
      app.find('Item').forEach((item) => {
        expect(item.props().onStatusChange).toEqual(app.instance().onStatusChange);
      });
    });
  });

  describe('Status copy', () => {
    beforeAll(() => {
      app.setState({
        items: [
          {
            id: 1,
            value: 'Item 1',
            isComplete: false,
          },
          {
            id: 2,
            value: 'Item 2',
            isComplete: true,
          },
          {
            id: 2,
            value: 'Item 2',
            isComplete: false,
          }
        ]
      })
    });

    it('should render updated copy for how many items are complete', () => {
      expect(app.find('.status').text()).toEqual('1 out of 3 tasks completed');
    });
  });

  describe('Task input', () => {
    it('should have two-way binding with a state property', () => {
      const inputField = app.find('#add-input');

      inputField.simulate('change', {
        target: {
          value: 'asdf'
        }
      });

      expect(app.state().newItemField).toEqual('asdf');
    });
  });

  describe('Adding items', () => {
    it('should add item to the state with the correct ID', ()=> {
      app.setState({
        newItemField: 'New item'
      });

      app.find('form').simulate('submit', {
        preventDefault: () => {}
      });

      expect(app.state().items.pop()).toEqual({
        id: app.state().items.length + 1,
        value: 'New item',
        isComplete: false
      });
    });

    it('should clear input when item is added', () => {
      app.setState({
        newItemField: 'New item'
      });

      app.find('form').simulate('submit', {
        preventDefault: () => {}
      });

      expect(app.state().newItemField).toEqual('');
    });
  });

  describe('Removing items', () => {
    beforeAll(() => {
      app.setState({
        items: [
          {
            id: 1,
            value: 'Item 1',
            isComplete: false,
          },
          {
            id: 2,
            value: 'Item 2',
            isComplete: false,
          }
        ]
      });
    })

    it('should remove item by ID', () => {
      app.instance().onRemove(1);

      expect(app.state().items).toHaveLength(1);
      expect(app.state().items[0].id).toEqual(2);
    });

    it('should pass remove method to each <Item /> component', () => {
      app.find('Item').forEach((itemComponent) => {
        expect(itemComponent.props().onRemove).toEqual(app.instance().onRemove);
      });
    });
  });
});
