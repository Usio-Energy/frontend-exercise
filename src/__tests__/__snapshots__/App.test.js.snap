// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Render should render correctly 1`] = `
<div>
  <h1>
    To do app
  </h1>
  <form
    onSubmit={[Function]}
  >
    <input
      id="add-input"
      onChange={[Function]}
      type="text"
      value=""
    />
    <input
      id="add-btn"
      type="submit"
      value="Add"
    />
  </form>
  <div
    className="status"
  >
    0
     out of 
    2
     tasks completed
  </div>
  <ul>
    <Item
      id={1}
      isComplete={false}
      onRemove={[Function]}
      onStatusChange={[Function]}
      value="Item 1"
    />
    <Item
      id={2}
      isComplete={false}
      onRemove={[Function]}
      onStatusChange={[Function]}
      value="Item 2"
    />
  </ul>
</div>
`;
